File: src/go/11.container-with-most-water.go
package main

/*
 * @lc app=leetcode id=11 lang=golang
 *
 * [11] Container With Most Water
 */

// @lc code=start
func maxArea(height []int) int {
	n := len(height)
	if n < 2 {
		return 0
	}
	l := 0
	r := n - 1
	maxarea := 0
	for l < r {
		maxarea = max(maxarea, min(height[l], height[r])*(r-l))
		if height[l] > height[r] {
			r--
		} else {
			l++
		}
	}
	return maxarea
}

// @lc code=end

File: src/java/two_pointer/11.container-with-most-water.java
class P11 {
    // @lc app=leetcode id=11 lang=java
    // @lc code=start
    public class Solution {
        public int maxArea(int[] height) {
            int i=0, j=height.length-1, best=0;
            while(i<j){
                int area = Math.min(height[i], height[j]) * (j-i);
                if (area>best) best=area;
                if (height[i] < height[j]) i++; else j--;
            }
            return best;
        }
    }
    // @lc code=end
}
